### PR TITLE
Added possibility to get asset state by index or state hash

### PR DESCRIPTION
**Pull request by u-hubar**

### CODE REVIEW

File: managers/asset-operations-manager.js [hunk 1]
```diff
@@ -344,21 +344,45 @@ class AssetOperationsManager {
...
-            ? await this.blockchainService.getUnfinalizedState(tokenId, blockchain)
-            : await this.blockchainService.getLatestAssertionId(tokenId, blockchain);
+        let publicAssertionId;
+        if (hasPendingUpdate) {
+            publicAssertionId = await this.blockchainService.getUnfinalizedState(tokenId, blockchain);
+        } else {
+            const assertionIds = await this.blockchainService.getAssertionIds(tokenId, blockchain);
+
+            if (state === ASSET_STATES.FINALIZED) {
+                publicAssertionId = assertionIds.slice(-1);
```
Comment (brkagithub) [inline managers/asset-operations-manager.js:360]: I might be misunderstanding the chart, but slice(-1) returns an array containing the last element and I think according to the chart we want just the last element. At least this line and line 369 should be in the same format (either element or subarray containing it) according to the chart

Comment (u-hubar) [reply managers/asset-operations-manager.js]: You're right, I fixed it

**Review by NZT48 — APPROVED** [review]

File: managers/asset-operations-manager.js [hunk 2]
```diff
@@ -344,21 +344,46 @@ class AssetOperationsManager {
...
-            ? await this.blockchainService.getUnfinalizedState(tokenId, blockchain)
-            : await this.blockchainService.getLatestAssertionId(tokenId, blockchain);
+        const unfinalizedState = await this.blockchainService.getUnfinalizedState(tokenId, blockchain);
+        const hasPendingUpdate = unfinalizedState !== ZeroHash;
+
+        let publicAssertionId;
+        let stateIsEnum = false;
+        if (hasPendingUpdate && ([ASSET_STATES.LATEST, unfinalizedState].includes(state))) {
+            publicAssertionId = unfinalizedState;
+            stateIsEnum = true;
```
Comment (zeroxbt) [inline managers/asset-operations-manager.js:355]: stateIsEnum should only be true here if state is === to ASSET_STATES.LATEST.

File: managers/asset-operations-manager.js [hunk 3]
```diff
@@ -344,21 +344,46 @@ class AssetOperationsManager {
...
+            }
+        }
 
         const getPublicOperationId = await this.nodeApiService.get(
             endpoint,
             port,
             authToken,
             UAL,
-            state,
+            stateIsEnum ? state: publicAssertionId,
```
Comment (zeroxbt) [inline managers/asset-operations-manager.js:386]: can't this just be Object.values(ASSET_STATES).includes(state) ? state : publicAssertionId

File: managers/asset-operations-manager.js [hunk 4]
```diff
@@ -344,21 +344,63 @@ class AssetOperationsManager {
...
+            (state === ASSET_STATES.LATEST && !isLatestUnfinalized)
+        ) {
+            publicAssertionId = assertionIds[assertionIds.length - 1];
+        } else if (typeof state === 'number') {
+            if (state >= assertionIds.length) {
+                throw Error('State index is out of range.');
+            }
+
+            publicAssertionId = assertionIds[state];
+        } else if (assertionIds.includes(state)) {
```
Comment (zeroxbt) [inline managers/asset-operations-manager.js:378]: assertionIds can be undefined here

Comment (u-hubar) [reply managers/asset-operations-manager.js:378]: Resolved

File: managers/asset-operations-manager.js [hunk 5]
```diff
@@ -346,53 +346,34 @@ class AssetOperationsManager {
         const { tokenId } = resolveUAL(UAL);
 
         let publicAssertionId;
-        let isLatestUnfinalized = false;
-        if (state === ASSET_STATES.LATEST) {
+        if(state === ASSET_STATES.LATEST || /^0x[a-fA-F0-9]{64}$/.test(state)) {
```
Comment (u-hubar) [inline managers/asset-operations-manager.js:349]: All hashes will pass this condition, even if it's a finalized state (we don't need to get pending state in this case)

**Review by djordjekovac — APPROVED** [review]

**Pull request merged**